File: flexui-android/flexui/src/main/kotlin/dev/flexui/FlexConfig.kt
package dev.flexui

import dev.flexui.schema.FlexTheme

/**
 * Configuration class for FlexUI SDK initialization
 */
class FlexConfig private constructor(
    val baseUrl: String,
    val cachePolicy: CachePolicy,
    val cacheTtlMs: Long,
    val defaultTheme: FlexTheme?,
    val connectTimeoutMs: Long,
    val readTimeoutMs: Long,
    val enableDebugLogging: Boolean
) {
    
    /**
     * Builder pattern for creating FlexConfig instances
     */
    class Builder(private val baseUrl: String) {
        private var cachePolicy: CachePolicy = CachePolicy.CACHE_FIRST
        private var cacheTtlMs: Long = 3600_000 // 1 hour
        private var defaultTheme: FlexTheme? = null
        private var connectTimeoutMs: Long = 10_000 // 10 seconds
        private var readTimeoutMs: Long = 30_000 // 30 seconds
        private var enableDebugLogging: Boolean = false
        
        /**
         * Set cache policy
         */
        fun cachePolicy(policy: CachePolicy): Builder {
            this.cachePolicy = policy
            return this
        }
        
        /**
         * Set cache TTL in milliseconds
         */
        fun cacheTtlMs(ttl: Long): Builder {
            this.cacheTtlMs = ttl
            return this
        }
        
        /**
         * Set cache TTL in seconds (convenience method)
         */
        fun cacheTtlSeconds(ttl: Long): Builder {
            this.cacheTtlMs = ttl * 1000
            return this
        }
        
        /**
         * Set cache TTL in minutes (convenience method)
         */
        fun cacheTtlMinutes(ttl: Long): Builder {
            this.cacheTtlMs = ttl * 60 * 1000
            return this
        }
        
        /**
         * Set default theme
         */
        fun defaultTheme(theme: FlexTheme): Builder {
            this.defaultTheme = theme
            return this
        }
        
        /**
         * Set connection timeout in milliseconds
         */
        fun connectTimeoutMs(timeout: Long): Builder {
            this.connectTimeoutMs = timeout
            return this
        }
        
        /**
         * Set connection timeout in seconds (convenience method)
         */
        fun connectTimeoutSeconds(timeout: Long): Builder {
            this.connectTimeoutMs = timeout * 1000
            return this
        }
        
        /**
         * Set read timeout in milliseconds
         */
        fun readTimeoutMs(timeout: Long): Builder {
            this.readTimeoutMs = timeout
            return this
        }
        
        /**
         * Set read timeout in seconds (convenience method)
         */
        fun readTimeoutSeconds(timeout: Long): Builder {
            this.readTimeoutMs = timeout * 1000
            return this
        }
        
        /**
         * Enable debug logging
         */
        fun enableDebugLogging(enable: Boolean = true): Builder {
            this.enableDebugLogging = enable
            return this
        }
        
        /**
         * Build the FlexConfig instance
         */
        fun build(): FlexConfig {
            return FlexConfig(
                baseUrl = baseUrl,
                cachePolicy = cachePolicy,
                cacheTtlMs = cacheTtlMs,
                defaultTheme = defaultTheme,
                connectTimeoutMs = connectTimeoutMs,
                readTimeoutMs = readTimeoutMs,
                enableDebugLogging = enableDebugLogging
            )
        }
    }
}

/**
 * Cache policy options for FlexUI configuration fetching
 */
enum class CachePolicy {
    /** Use cache if available, fetch in background to update */
    CACHE_FIRST,
    
    /** Always fetch from network first, fallback to cache on failure */
    NETWORK_FIRST,
    
    /** Only use cache, never fetch from network (offline mode) */
    CACHE_ONLY,
    
    /** Always fetch from network, never use cache */
    NETWORK_ONLY
}

File: flexui-ios/Sources/FlexUI/FlexConfig.swift
import Foundation

/// Configuration object for FlexUI initialization
public struct FlexConfig {
    
    // MARK: - Properties
    
    /// Base URL for fetching configurations
    public let baseURL: String
    
    /// Cache policy for configuration fetching
    public let cachePolicy: CachePolicy
    
    /// Cache time-to-live in seconds
    public let cacheTTL: TimeInterval
    
    /// Default theme to merge with server-provided themes
    public let defaultTheme: FlexTheme?
    
    /// Connection timeout for network requests
    public let connectTimeout: TimeInterval
    
    /// Read timeout for network requests
    public let readTimeout: TimeInterval
    
    // MARK: - Initializers
    
    /// Initialize with base URL and default settings
    public init(baseURL: String) {
        self.baseURL = baseURL
        self.cachePolicy = .cacheFirst
        self.cacheTTL = 3600 // 1 hour
        self.defaultTheme = nil
        self.connectTimeout = 10
        self.readTimeout = 30
    }
    
    /// Initialize with all configuration options
    public init(
        baseURL: String,
        cachePolicy: CachePolicy = .cacheFirst,
        cacheTTL: TimeInterval = 3600,
        defaultTheme: FlexTheme? = nil,
        connectTimeout: TimeInterval = 10,
        readTimeout: TimeInterval = 30
    ) {
        self.baseURL = baseURL
        self.cachePolicy = cachePolicy
        self.cacheTTL = cacheTTL
        self.defaultTheme = defaultTheme
        self.connectTimeout = connectTimeout
        self.readTimeout = readTimeout
    }
}

// MARK: - Cache Policy

/// Defines how FlexUI handles configuration caching
public enum CachePolicy {
    /// Use cache if available, fetch in background to update
    case cacheFirst
    
    /// Always fetch from network first, fallback to cache on failure
    case networkFirst
    
    /// Never fetch from network, only use cached configurations
    case cacheOnly
    
    /// Never use cache, always fetch fresh from network
    case networkOnly
}

// MARK: - Convenience Builder

public extension FlexConfig {
    
    /// Builder pattern for creating FlexConfig with optional parameters
    struct Builder {
        private var baseURL: String
        private var cachePolicy: CachePolicy = .cacheFirst
        private var cacheTTL: TimeInterval = 3600
        private var defaultTheme: FlexTheme?
        private var connectTimeout: TimeInterval = 10
        private var readTimeout: TimeInterval = 30
        
        /// Initialize builder with base URL
        public init(baseURL: String) {
            self.baseURL = baseURL
        }
        
        /// Set cache policy
        public func cachePolicy(_ policy: CachePolicy) -> Builder {
            var builder = self
            builder.cachePolicy = policy
            return builder
        }
        
        /// Set cache TTL in seconds
        public func cacheTTL(_ ttl: TimeInterval) -> Builder {
            var builder = self
            builder.cacheTTL = ttl
            return builder
        }
        
        /// Set default theme
        public func defaultTheme(_ theme: FlexTheme) -> Builder {
            var builder = self
            builder.defaultTheme = theme
            return builder
        }
        
        /// Set connection timeout in seconds
        public func connectTimeout(_ timeout: TimeInterval) -> Builder {
            var builder = self
            builder.connectTimeout = timeout
            return builder
        }
        
        /// Set read timeout in seconds
        public func readTimeout(_ timeout: TimeInterval) -> Builder {
            var builder = self
            builder.readTimeout = timeout
            return builder
        }
        
        /// Build the final FlexConfig
        public func build() -> FlexConfig {
            return FlexConfig(
                baseURL: baseURL,
                cachePolicy: cachePolicy,
                cacheTTL: cacheTTL,
                defaultTheme: defaultTheme,
                connectTimeout: connectTimeout,
                readTimeout: readTimeout
            )
        }
    }
}

// MARK: - Convenience Methods

public extension FlexConfig {
    
    /// Create a FlexConfig with cache-first policy and custom TTL
    static func cacheFirst(baseURL: String, cacheTTL: TimeInterval = 3600) -> FlexConfig {
        return FlexConfig(baseURL: baseURL, cachePolicy: .cacheFirst, cacheTTL: cacheTTL)
    }
    
    /// Create a FlexConfig with network-first policy
    static func networkFirst(baseURL: String) -> FlexConfig {
        return FlexConfig(baseURL: baseURL, cachePolicy: .networkFirst)
    }
    
    /// Create a FlexConfig with cache-only policy (offline mode)
    static func cacheOnly(baseURL: String) -> FlexConfig {
        return FlexConfig(baseURL: baseURL, cachePolicy: .cacheOnly)
    }
    
    /// Create a FlexConfig with network-only policy (no caching)
    static func networkOnly(baseURL: String) -> FlexConfig {
        return FlexConfig(baseURL: baseURL, cachePolicy: .networkOnly)
    }
}